The code base fails to build with NONMATCHING defined

If I add `#define NONMATCHING 1` to include/global.h and try to build, a bunch of errors will show up: 
```
src/bmlib.c: In function `sub_8013278':
src/bmlib.c:381: warning: assignment from incompatible pointer type
src/bmlib.c:382: warning: assignment from incompatible pointer type
src/bmlib.c:384: warning: assignment from incompatible pointer type
src/bmlib.c:385: warning: assignment from incompatible pointer type
src/bmlib.c:408: `gGenericBuffer16' undeclared (first use in this function)
src/bmlib.c:408: (Each undeclared identifier is reported only once
src/bmlib.c:408: for each function it appears in.)

src/bmmind.c:707: syntax error before `?'
src/bmmind.c:708: syntax error before `('

......
```

--- a/include/bmsave.h
+++ b/include/bmsave.h
@@ -6,6 +6,7 @@
 #include "bmbattle.h"
 #include "bmtrick.h"
 #include "bmdifficulty.h"
+#include "bonusclaim.h"
 
 enum {
     UNIT_SAVE_AMOUNT_BLUE = 51,
@@ -134,6 +135,11 @@ struct bmsave_unkstruct2 {
     u32 unk[0x4];
     u16 magic1;
     u16 magic2;
+};
+
+struct bmsave_unkstruct3 {
+    struct BonusClaimEnt unk0[0x10];
+    u32 unk140;
 };
 
 enum
@@ -356,13 +362,13 @@ struct GameSaveBlock {
     u32 bonusClaimFlags;
     struct GMapSaveInfo wmStuff;
     struct Dungeon dungeons[2];
-};
+}; /* size = 0xDC8 */
 
 struct SuspendSaveBlock {
     struct PlaySt playSt;
     struct ActionData action;
     struct SuspendSavePackedUnit blueUnits[UNIT_SAVE_AMOUNT_BLUE];
-    struct SuspendSavePackedUnit wmMonsterUnit;
+    struct SuspendSavePackedUnit wmMonsterUnit; // TODO: update this to `struct Dungeon dungeons[2]; u8 filler[0x1C];`?
     struct SuspendSavePackedUnit redUnits[UNIT_SAVE_AMOUNT_RED];
     struct SuspendSavePackedUnit greenUnits[UNIT_SAVE_AMOUNT_GREEN];
     struct Trap traps[TRAP_MAX_COUNT];
@@ -375,6 +381,44 @@ struct SuspendSaveBlock {
     struct GMapSaveInfo wmStuff;
     struct Dungeon dungeon;
     int eventSlotCnt;
+}; /* size = 0x1F78 */
+
+struct SaveBlocks {
+    /* 0x0000 */ struct GlobalSaveInfo globalSaveInfo;
+    /* 0x0064 */ struct SaveBlockInfo saveBlockInfo[SAVE_ID_MAX];
+    /* 0x00D4 */ struct SuspendSaveBlock suspendSaveBlocks[2];
+    /* 0x3FC4 */ struct GameSaveBlock gameSaveBlocks[3];
+    /* 0x691C */ u8 offset_5[0x874];
+    // TODO: update these fields
+    /* 0x7190 */ u8 unk_7190[0x270];
+    /* 0x7400 */ struct ExtraMapSaveHead offset_6; // see bmsave-xmap.c
+};
+
+// TODO: figure out how these structs work
+struct GameSaveBlockEwram {
+    struct PlaySt playSt;
+    u8 unk_4C[0xC];
+    u8 buffer_58[0xD1C];
+    u32 buffer_D74[0xC];
+}; /* size = 0xDA4 */
+
+struct SuspendSaveBlockEwram {
+    struct PlaySt playSt;
+    u8 unk_4C[0xC];
+    u8 buffer_58[0x1EC4];
+    u32 buffer_1F1C[0xF];
+}; /* size = 0x1F58 */
+
+struct SaveBlocksEwram {
+    /* 0x0000 */ struct GlobalSaveInfo globalSaveInfo;
+    /* 0x0064 */ struct SaveBlockInfo saveBlockInfo[SAVE_ID_MAX];
+    /* 0x00D4 */ struct SuspendSaveBlockEwram suspendSaveBlocks[2];
+    /* 0x3F84 */ struct GameSaveBlockEwram gameSaveBlocks[3];
+    /* 0x6870 */ u8 offset_5[0x874];
+    /* 0x70E4 */ struct GameRankSaveDataPacks gameRankSave;
+    /* 0x7178 */ struct bmsave_unkstruct1 unkstruct1;
+    /* 0x719C */ struct bmsave_unkstruct2 unkstruct2;
+    /* 0x71B0 */ struct bmsave_unkstruct3 unkstruct3;
 };
 
 extern struct UnitUsageStats *gPidStatsSaveLoc;
@@ -414,7 +458,7 @@ void WriteChapterFlags(void *sram_dest);
 void WritePermanentFlags(void *sram_dest);
 void ReadChapterFlags(void *ewram_dest);
 void ReadPermanentFlags(void *ewram_dest);
-void ReadPermanentFlags_ret(void *sram_src, void *ewram_dest);
+void ReadPermanentFlags_ret(const void *sram_src, void *ewram_dest);
 void WriteSupplyItems(void *sram_dest);
 void ReadSupplyItems(const void *sram_src);
 bool null_true(void);
@@ -459,7 +503,7 @@ void sub_80A4000(struct bmsave_unkstruct2 *buf);
 int sub_80A402C(void *buf, int val);
 int sub_80A402C(void *buf, int val);
 void LoadAndVerifySramSaveData(void);
-void ClearPidChStatsSaveData(void *sram_dest);
+void ClearPidChStatsSaveData(struct GameSaveBlock *sram_dest);
 void ClearPidStats_ret(void);
 void ClearPidStats(void);
 void ReadPidStats(void *sram_src);
@@ -510,8 +554,8 @@ void SetGameEndFlag(void);
 struct UnitUsageStats* GetPidStats(u8 pid);
 u32 GetBonusContentClaimFlags(void);
 void SetBonusContentClaimFlags(u32 num);
-void WriteBonusContentClaimFlags(void *sram_dest);
-void ReadBonusContentClaimFlags(const void *sram_src);
+void WriteBonusContentClaimFlags(struct GameSaveBlock *sram_dest);
+void ReadBonusContentClaimFlags(const struct GameSaveBlock *sram_src);
 void WriteLastGameSaveId(int num);
 int ReadLastGameSaveId(void);
 void InvalidateGameSave(int);
@@ -522,7 +566,7 @@ void ReadGameSave(int slot);
 bool IsSaveValid(int);
 void ReadGameSavePlaySt(int, struct PlaySt*);
 u32 LoadSavedBonusClaimFlags(int slot);
-void LoadSavedWMStuff(int slot, void *dest);
+void LoadSavedWMStuff(int slot, struct GMapData *dest);
 s8 LoadSavedEid8A(int slot);
 bool IsGameNotFirstChapter(struct PlaySt *chapter_data);
 bool IsGameSaveNotFirstChapter(int);

--- a/include/functions.h
+++ b/include/functions.h
@@ -2619,7 +2619,7 @@ char* sub_80BBA28(int); // GetWorldMapNodeName
 // ??? sub_80BC7F4(???);
 void sub_80BC81C(void);
 // ??? sub_80BC898(???);
-bool SetupNewWMRoute(struct GMapData*, u8*, int);
+bool SetupNewWMRoute(struct GMapData*, struct OpenPaths*, int);
 // ??? WM_RemovePath(???);
 // ??? sub_80BC970(???);
 void sub_80BCA0C(struct GMapData*);

--- a/include/types.h
+++ b/include/types.h
@@ -181,6 +181,26 @@ enum BmSt_gameStateBits {
     BM_FLAG_LINKARENA = (1 << 6),
 };
 
+struct PlaySt_30 {
+    int total_gold;
+
+    u32 unk_4_00 : 0x14;
+    u32 unk_4_14 : 0x0C;
+
+    u32 unk_8_1:8;
+    u32 unk_8_2:20; // Used by bmdifficulty (Valni/Lagdou)
+    u32 unk_8_3:4;
+
+    u32 unk_C_00 : 6;
+    u32 combatRank : 3;
+    u32 expRank : 3;
+    u32 unk_3D_04 : 3;
+    u32 fundsRank : 3;
+    u32 tacticsRank : 3;
+    u32 survivalRank : 3;
+    u32 unk_F_00 : 8;
+};
+
 struct PlaySt { // Chapter Data Struct
     /* 00 */ u32 time_saved;
     /* 04 */ u32 time_chapter_started;
@@ -225,23 +245,7 @@ struct PlaySt { // Chapter Data Struct
     u32 unk_2C_2:5;
     u32 unk_2C_3:4;
 
-    /* 30 */ int total_gold;
-
-    /* 34 */ u32 unk_34_00 : 0x14;
-             u32 unk_34_14 : 0x0C;
-
-    u32 unk_38_1:8;
-    u32 unk_38_2:20; // Used by bmdifficulty (Valni/Lagdou)
-    u32 unk_38_3:4;
-
-    /* 3C */ u32 unk_3C_00 : 6;
-    /* 3C */ u32 combatRank : 3;
-    /* 3D */ u32 expRank : 3;
-    /* 3D */ u32 unk_3D_04 : 3;
-    /* 3D */ u32 fundsRank : 3;
-    /* 3E */ u32 tacticsRank : 3;
-    /* 3E */ u32 survivalRank : 3;
-    /* 3F */ u32 unk_3F_00 : 8;
+    struct PlaySt_30 unk_30;
 
     // option bits
     u32 cfgUnitColor:1; // 1

--- a/include/types.h
+++ b/include/types.h
@@ -493,6 +493,11 @@ union GMStateBits {
     } __attribute__((packed)) bits;
 } __attribute__((packed));
 
+struct OpenPaths {
+    s8 openPaths[0x20];
+    s8 openPathsLength;
+};
+
 struct GMapData
 {
     /* 00 */ union GMStateBits state;
@@ -504,11 +509,7 @@ struct GMapData
     /* 10 */ struct GMUnit unk10[8];
     /* 30 */ struct GMNode unk30[0x1C];
     /* A0 */ int unk_a0; // pad?
-    // These are typically treated as one logical unit, but can't be pulled out
-    // into a separate struct because that messes up the struct padding.
-    /* A4 */ s8 openPaths[0x20];
-    /* C4 */ s8 openPathsLength;
-    /* C5 */ u8 _pad[3];
+    /* A4 */ struct OpenPaths routeData;
     /* C8 */ u8 unk_c8; // entry node id?
     /* C9 */ u8 unk_c9[3]; // List of active world map skirmishes
     /* CC */ u8 unk_cc; // used to determine which skirmish enemy block to load

--- a/src/bmdifficulty.c
+++ b/src/bmdifficulty.c
@@ -287,7 +287,7 @@ void UpdateDungeonStats(struct Dungeon* dungeon) {
 
     val = dungeon->expEarned;
 
-    val += (PidStatsGetTotalExpGain() - gPlaySt.unk_38_2);
+    val += (PidStatsGetTotalExpGain() - gPlaySt.unk_30.unk_8_2);
 
     if (val > 50000) {
         val = 50000;

--- a/src/bmsave-bwl.c
+++ b/src/bmsave-bwl.c
@@ -11,7 +11,7 @@ EWRAM_DATA struct UnitUsageStats *gPidStatsSaveLoc = NULL;
 EWRAM_DATA struct UnitUsageStats gPidStatsData[BWL_ARRAY_NUM] = {0};
 EWRAM_DATA struct ChapterStats gChapterStats[WIN_ARRAY_NUM] = {0};
 
-void ClearPidChStatsSaveData(void *sram_dest)
+void ClearPidChStatsSaveData(struct GameSaveBlock *sram_dest)
 {
     int i, j;
 
@@ -20,13 +20,13 @@ void ClearPidChStatsSaveData(void *sram_dest)
 
     for (i = 0; i < BWL_ARRAY_NUM; i++) {
         gPidStatsData[i].favval = 0x2000;
-        WriteAndVerifySramFast(gPidStatsData, sram_dest + GAMESAVE_OFFSET_PIDSTATS + i * sizeof(struct UnitUsageStats), sizeof(struct UnitUsageStats));
+        WriteAndVerifySramFast(gPidStatsData, &sram_dest->pidStats[i], sizeof(struct UnitUsageStats));
     }
 
     for (i = 0; i < WIN_ARRAY_NUM; i++)
-        WriteAndVerifySramFast(gChapterStats, sram_dest + GAMESAVE_OFFSET_CHAPTERSTATS + i * sizeof(struct ChapterStats), sizeof(struct ChapterStats));
+        WriteAndVerifySramFast(gChapterStats, &sram_dest->chapterStats[i], sizeof(struct ChapterStats));
 
-    gPidStatsSaveLoc = sram_dest + 0x084C;
+    gPidStatsSaveLoc = sram_dest->pidStats;
 }
 
 void ClearPidStats_ret()
@@ -39,11 +39,11 @@ void ClearPidStats_ret()
 void ClearPidStats()
 {
     CpuFill16(0, gPidStatsData, sizeof(gPidStatsData));
-    gPlaySt.unk_38_2 = 0;
-    gPlaySt.unk_34_14 = 0;
-    gPlaySt.unk_38_1 = 0;
-    gPlaySt.unk_34_00 = 0;
-    gPlaySt.total_gold = GetPartyTotalGoldValue();
+    gPlaySt.unk_30.unk_8_2 = 0;
+    gPlaySt.unk_30.unk_4_14 = 0;
+    gPlaySt.unk_30.unk_8_1 = 0;
+    gPlaySt.unk_30.unk_4_00 = 0;
+    gPlaySt.unk_30.total_gold = GetPartyTotalGoldValue();
 }
 
 void ReadPidStats(void *sram_src)
@@ -124,8 +124,8 @@ void RegisterChapterTimeAndTurnCount(struct PlaySt* play_st)
         time = 50 * 60 * 20;
 
     turn = play_st->chapterTurnNumber;
-    if (turn > 0x1F4)
-        turn = 0x1F4;
+    if (turn > 500)
+        turn = 500;
 
     chstat->chapter_index = play_st->chapterIndex;
     chstat->chapter_turn = turn;
@@ -250,7 +250,7 @@ void PidStatsAddWinAmt(u8 pid)
     if (NULL == bwl)
         return;
 
-    if (bwl->winAmt < 0x3E8)
+    if (bwl->winAmt < 1000)
         bwl->winAmt++;
 
     PidStatsAddFavval(pid, 0x10);

--- a/src/bmsave-gmap.c
+++ b/src/bmsave-gmap.c
@@ -86,8 +86,8 @@ void WriteWorldMapPaths(struct GMapData* pGMapData, u8* pathFlags) {
         pathFlags[i] = 0;
     }
 
-    for (i = 0; i < pGMapData->openPathsLength; i++) {
-        int path = pGMapData->openPaths[i];
+    for (i = 0; i < pGMapData->routeData.openPathsLength; i++) {
+        int path = pGMapData->routeData.openPaths[i];
 
         int idx = path / 8;
         u8* addr = pathFlags + idx;
@@ -104,13 +104,13 @@ void WriteWorldMapPaths(struct GMapData* pGMapData, u8* pathFlags) {
 void ReadWorldMapPaths(struct GMapData* pGMapData, u8* pathFlags) {
     int i;
 
-    pGMapData->openPathsLength = 0;
+    pGMapData->routeData.openPathsLength = 0;
 
     for (i = 0; i < 0x20; i++) {
         int idx = i / 8;
 
         if ((pathFlags[idx] >> (i & 7)) & 1) {
-            SetupNewWMRoute(pGMapData, pGMapData->openPaths, i);
+            SetupNewWMRoute(pGMapData, &pGMapData->routeData, i);
         }
     }
 

--- a/src/bmsave-lib.c
+++ b/src/bmsave-lib.c
@@ -246,7 +246,7 @@ void WriteSaveBlockInfo(struct SaveBlockInfo *chunk, int index)
 
     switch (chunk->kind) {
     case SAVEBLOCK_KIND_GAME:
-        chunk->size = GAMESAVE_OFFSET_MAX;
+        chunk->size = sizeof(struct GameSaveBlock);
         break;
 
     case SAVEBLOCK_KIND_SUSPEND:
@@ -364,7 +364,7 @@ void ReadPermanentFlags(void *ewram_dest)
         GetPermanentFlagBitsSize());
 }
 
-void ReadPermanentFlags_ret(void *sram_src, void *ewram_dest)
+void ReadPermanentFlags_ret(const void *sram_src, void *ewram_dest)
 {
     ReadSramFast(
         sram_src,

--- a/src/bmsave-misc.c
+++ b/src/bmsave-misc.c
@@ -42,35 +42,31 @@ void sub_80A7258(struct PlaySt* src, struct PlaySt* dst) {
     return;
 }
 
-struct UnkStruct_SubA7298 {
-    u8 unk[0x1d];
-};
-
 //! FE8U = 0x080A7298
-void sub_80A7298(struct UnkStruct_SubA7298* src, struct UnkStruct_SubA7298* dst) {
-    memcpy(dst, src, 0x1d);
+void sub_80A7298(const u32* src, u32* dst) {
+    memcpy(dst, src, 0x1d); // TODO: where does the 0x1D come from?
 }
 
 //! FE8U = 0x080A72B0
-void sub_80A72B0(struct SuspendSaveBlock* src, struct SuspendSaveBlock* dst) {
-    memcpy((void*)dst + 0x4c, (void*)src + 0x58, 0x00001EC4);
+void sub_80A72B0(struct SuspendSaveBlockEwram* src, struct SuspendSaveBlock* dst) {
+    memcpy(&dst->action, src->buffer_58, sizeof(src->buffer_58));
     sub_80A7258(&src->playSt, &dst->playSt);
-    sub_80A7298((struct UnkStruct_SubA7298*)(src->menuOverride + 8), (struct UnkStruct_SubA7298*)(dst->menuOverride + 8 + 0x28));
+    sub_80A7298(src->buffer_1F1C, (void*)&dst->wmStuff);
 
     return;
 }
 
 //! FE8U = 0x080A72EC
-void sub_80A72EC(struct GameSaveBlock* src, struct GameSaveBlock* dst) {
-    memcpy((void*)dst + 0x4c, (void*)src + 0x58, 0x00000D1C);
+void sub_80A72EC(struct GameSaveBlockEwram* src, struct GameSaveBlock* dst) {
+    memcpy(dst->units, src->buffer_58, sizeof(src->buffer_58));
     sub_80A7258(&src->playSt, &dst->playSt);
-    sub_80A7298((struct UnkStruct_SubA7298*)(src->permanentFlags + 8), (struct UnkStruct_SubA7298*)(dst->permanentFlags + 8 + 0x18));
+    sub_80A7298(src->buffer_D74, (void*)&dst->wmStuff);
     return;
 }
 
 //! FE8U = 0x080A7328
-void sub_80A7328(struct BonusClaimEnt* src, struct BonusClaimEnt* dst) {
-    memcpy(dst, src, sizeof(struct BonusClaimEnt) * 16 + 4);
+void sub_80A7328(struct bmsave_unkstruct3* src, struct bmsave_unkstruct3* dst) {
+    *dst = *src;
     return;
 }
 
@@ -98,16 +94,16 @@ extern u8 gUnknown_02008000[];
 //! FE8U = 0x080A7374
 void sub_80A7374(void) {
     int i;
-    void* ewram;
-    void* sram;
-    void* ewram_;
+    struct SaveBlocksEwram* ewram;
+    struct SaveBlocks* sram;
+    struct SaveBlocksEwram* ewram_;
 
-    ewram = gAnims;
+    ewram = (void*)gAnims;
     sram = (void*)0xe000000;
 
-    ReadSramFast(sram, ewram, 0x64);
+    ReadSramFast(&sram->globalSaveInfo, ewram, sizeof(struct GlobalSaveInfo));
 
-    if (((struct GlobalSaveInfo*)(ewram))->magic32 == 0x00040624) {
+    if (ewram->globalSaveInfo.magic32 == 0x00040624) {
         return;
     }
 
@@ -115,48 +111,48 @@ void sub_80A7374(void) {
 
     ReadSramFast(sram, ewram_, 0x8000);
 
-    if (((struct GlobalSaveInfo*)(ewram_))->magic32 != 0x00040603) {
+    if (ewram_->globalSaveInfo.magic32 != 0x00040603) {
         return;
     }
 
-    CpuFastFill(0, ewram + 0x8000, sizeof(struct GlobalSaveInfo));
-    sub_80A720C(ewram_, ewram + 0x8000);
-    WriteGlobalSaveInfo((struct GlobalSaveInfo*)(ewram + 0x8000));
+    CpuFastFill(0, (void *)ewram + 0x8000, sizeof(struct GlobalSaveInfo));
+    sub_80A720C(&ewram_->globalSaveInfo, &((struct SaveBlocksEwram*)((void *)ewram + 0x8000))->globalSaveInfo);
+    WriteGlobalSaveInfo(&((struct SaveBlocksEwram*)((void *)ewram + 0x8000))->globalSaveInfo);
 
     for (i = 0; i < 2; i++) {
-        CpuFastFill(0, ewram + 0x8000, sizeof(struct SuspendSaveBlock));
-        sub_80A72B0(ewram_ + (i * 0x00001F58 + 0xd4), ewram + 0x8000);
-        WriteAndVerifySramFast(ewram + 0x8000, sram + 0xd4 + i * sizeof(struct SuspendSaveBlock), sizeof(struct SuspendSaveBlock));
+        CpuFastFill(0, (void *)ewram + 0x8000, sizeof(struct SuspendSaveBlock));
+        sub_80A72B0(&ewram_->suspendSaveBlocks[i], (void *)ewram + 0x8000);
+        WriteAndVerifySramFast((void *)ewram + 0x8000, &sram->suspendSaveBlocks[i], sizeof(struct SuspendSaveBlock));
     }
 
     for (i = 0; i < 3; i++) {
         CpuFastFill(0, gUnknown_02008000, sizeof(struct GameSaveBlock));
-        sub_80A72EC(ewram_ + (i * 0x00000DA4 + 0x00003F84), (struct GameSaveBlock*)gUnknown_02008000);
-        WriteAndVerifySramFast(gUnknown_02008000, (sram + 0x00003FC4) + (i * sizeof(struct GameSaveBlock)), sizeof(struct GameSaveBlock));
+        sub_80A72EC(&ewram_->gameSaveBlocks[i], (struct GameSaveBlock*)gUnknown_02008000);
+        WriteAndVerifySramFast(gUnknown_02008000, &sram->gameSaveBlocks[i], sizeof(struct GameSaveBlock));
     }
 
     // GameRankSaveDataPacks
-    CpuFastFill(0, gUnknown_02008000, 0x94);
-    sub_80A7360(ewram_ + 0x000070E4, (struct GameRankSaveDataPacks*)gUnknown_02008000);
+    CpuFastFill(0, gUnknown_02008000, sizeof(struct GameRankSaveDataPacks));
+    sub_80A7360(&ewram_->gameRankSave, (struct GameRankSaveDataPacks*)gUnknown_02008000);
     SaveRankings(gUnknown_02008000);
 
     // "bmsave_unkstruct1", flags for sound room data?
     CpuFastFill(0, gUnknown_02008000, sizeof(struct bmsave_unkstruct1));
-    sub_80A734C(ewram_ + 0x00007178, (struct bmsave_unkstruct1*)gUnknown_02008000);
+    sub_80A734C(&ewram_->unkstruct1, (struct bmsave_unkstruct1*)gUnknown_02008000);
     sub_80A3EA4(gUnknown_02008000);
 
     // "bmsave_unkstruct2", flags for viewing CGs?
     CpuFastFill(0, gUnknown_02008000, sizeof(struct bmsave_unkstruct2));
-    sub_80A733C(ewram_ + 0x0000719C, (struct bmsave_unkstruct2*)gUnknown_02008000);
+    sub_80A733C(&ewram_->unkstruct2, (struct bmsave_unkstruct2*)gUnknown_02008000);
     sub_80A4000((struct bmsave_unkstruct2*)gUnknown_02008000);
 
-    CpuFastFill(0, gUnknown_02008000, sizeof(struct BonusClaimEnt) * 16 + 4);
-    sub_80A7328(ewram_ + 0x000071B0, (struct BonusClaimEnt*)gUnknown_02008000);
+    CpuFastFill(0, gUnknown_02008000, sizeof(struct bmsave_unkstruct3));
+    sub_80A7328(&ewram_->unkstruct3, (struct bmsave_unkstruct3*)gUnknown_02008000);
     SaveBonusContentData(gUnknown_02008000);
 
-    for (i = 0; i < 7; i++) {
+    for (i = 0; i < SAVE_ID_MAX; i++) {
         CpuFastFill(0, gUnknown_02008000, sizeof(struct SaveBlockInfo));
-        sub_80A723C((ewram_ + 0x64) + i * sizeof(struct SaveBlockInfo), (struct SaveBlockInfo*)gUnknown_02008000);
+        sub_80A723C(&ewram_->saveBlockInfo[i], (struct SaveBlockInfo*)gUnknown_02008000);
         WriteSaveBlockInfo((struct SaveBlockInfo *)gUnknown_02008000, i);
     }
 

--- a/src/bmsave.c
+++ b/src/bmsave.c
@@ -31,16 +31,16 @@ void SetBonusContentClaimFlags(u32 num)
     gBonusContentClaimFlags = num;
 }
 
-void WriteBonusContentClaimFlags(void *sram_dest)
+void WriteBonusContentClaimFlags(struct GameSaveBlock *sram_dest)
 {
     WriteAndVerifySramFast(&gBonusContentClaimFlags,
-                           sram_dest + GAMESAVE_OFFSET_BONUSCLAIM,
+                           &sram_dest->bonusClaimFlags,
                            sizeof(gBonusContentClaimFlags));
 }
 
-void ReadBonusContentClaimFlags(const void *sram_src)
+void ReadBonusContentClaimFlags(const struct GameSaveBlock *sram_src)
 {
-    ReadSramFast(sram_src + GAMESAVE_OFFSET_BONUSCLAIM,
+    ReadSramFast(&sram_src->bonusClaimFlags,
                  &gBonusContentClaimFlags,
                  sizeof(gBonusContentClaimFlags));
 }
@@ -94,8 +94,8 @@ void CopyGameSave(int index_src, int index_dest)
     void *src = GetSaveReadAddr(index_src);
     void *dest = GetSaveWriteAddr(index_dest);
 
-    ReadSramFast(src, gGenericBuffer, 0xDC8);
-    WriteAndVerifySramFast(gGenericBuffer, dest, 0xDC8);
+    ReadSramFast(src, gGenericBuffer, sizeof(struct GameSaveBlock));
+    WriteAndVerifySramFast(gGenericBuffer, dest, sizeof(struct GameSaveBlock));
 
     chunk.magic32 = SAVEMAGIC32;
     chunk.kind = SAVEBLOCK_KIND_GAME;
@@ -108,8 +108,7 @@ void WriteNewGameSave(int index, int isDifficult, int mode, int isTutorial)
     struct SaveBlockInfo chunk;
     struct GameSavePackedUnit unitp;
 
-    void *dest = GetSaveWriteAddr(index);
-    void *tmp_dest;
+    struct GameSaveBlock *dest = GetSaveWriteAddr(index);
 
     if (0 == mode)
         mode = gPlaySt.chapterModeIndex;
@@ -126,7 +125,7 @@ void WriteNewGameSave(int index, int isDifficult, int mode, int isTutorial)
     
     gPlaySt.unk_2C_1 = 0;
     gPlaySt.unk_2C_04 = 0;
-    CpuFill16(0, &gPlaySt.total_gold, 0x10);
+    CpuFill16(0, &gPlaySt.unk_30, sizeof(gPlaySt.unk_30));
     gPlaySt.unk_2C_0D = 0;
     gPlaySt.chapterModeIndex = mode;
     gPlaySt.unk_2B_00 = 1;
@@ -136,26 +135,26 @@ void WriteNewGameSave(int index, int isDifficult, int mode, int isTutorial)
     gPlaySt.gameSaveSlot = index;
     gPlaySt.unk_2C_2 = GetGlobalCompletionCount();
 
-    WriteAndVerifySramFast(&gPlaySt, dest + GAMESAVE_OFFSET_PLAYST, sizeof(gPlaySt));
+    WriteAndVerifySramFast(&gPlaySt, &dest->playSt, sizeof(gPlaySt));
     SetBonusContentClaimFlags(0);
     WriteBonusContentClaimFlags(dest);
     
     CpuFill16(0, &unitp, sizeof(unitp));
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++) {
-        WriteAndVerifySramFast(&unitp, dest + GAMESAVE_OFFSET_UNITS + i * sizeof(unitp), sizeof(unitp));
+        WriteAndVerifySramFast(&unitp, &dest->units[i], sizeof(unitp));
     }
-    WriteAndVerifySramFast(&unitp, dest + GAMESAVE_OFFSET_GMMONSTER, sizeof(unitp));
+    WriteAndVerifySramFast(&unitp, &dest->gmUnit, sizeof(unitp));
 
-    WriteSupplyItems(dest + GAMESAVE_OFFSET_SUPPLY);
+    WriteSupplyItems(dest->supplyItems);
     ClearPidChStatsSaveData(dest);
-    WritePermanentFlags(dest + GAMESAVE_OFFSET_PERMANENTFLAGS);
-    ClearWorldMapStuff(dest + GAMESAVE_OFFSET_GMINFO);
+    WritePermanentFlags(dest->permanentFlags);
+    ClearWorldMapStuff(&dest->wmStuff);
 
     {
         struct Dungeon dungeon[2];  
         CpuFill16(0, dungeon, sizeof(dungeon));
-        WriteAndVerifySramFast(dungeon, dest + GAMESAVE_OFFSET_DUNGEON, sizeof(dungeon));
+        WriteAndVerifySramFast(dungeon, dest->dungeons, sizeof(dungeon));
     }
 
     chunk.magic32 = SAVEMAGIC32;
@@ -171,14 +170,14 @@ void WriteGameSave(int slot)
     struct GlobalSaveInfo info;
     struct Dungeon dungeon[2];
 
-    void *dest = GetSaveWriteAddr(slot);
+    struct GameSaveBlock *dest = GetSaveWriteAddr(slot);
     InvalidateSuspendSave(SAVE_ID_SUSPEND);
     gPlaySt.gameSaveSlot = slot;
     gPlaySt.time_saved = GetGameClock();
-    WriteAndVerifySramFast(&gPlaySt, dest, sizeof(gPlaySt));
+    WriteAndVerifySramFast(&gPlaySt, &dest->playSt, sizeof(gPlaySt));
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++)
-        WriteGameSavePackedUnit(&gUnitArrayBlue[i], dest + 0x4C + i * sizeof(struct GameSavePackedUnit));
+        WriteGameSavePackedUnit(&gUnitArrayBlue[i], &dest->units[i]);
 
     ReadGlobalSaveInfo(&info);
     
@@ -186,15 +185,15 @@ void WriteGameSave(int slot)
         SGM_SetCharacterKnown(UNIT_CHAR_ID(&gUnitArrayBlue[i]), &info);
 
     WriteGlobalSaveInfo(&info);
-    WriteGMMonsterRnState(dest + GAMESAVE_OFFSET_GMMONSTER);
-    WriteSupplyItems(dest + GAMESAVE_OFFSET_SUPPLY);
-    WritePidStats(dest + GAMESAVE_OFFSET_PIDSTATS);
-    WriteChapterStats(dest + GAMESAVE_OFFSET_CHAPTERSTATS);
+    WriteGMMonsterRnState(&dest->gmUnit);
+    WriteSupplyItems(dest->supplyItems);
+    WritePidStats(dest->pidStats);
+    WriteChapterStats(dest->chapterStats);
     WriteBonusContentClaimFlags(dest);
-    WritePermanentFlags(dest + GAMESAVE_OFFSET_PERMANENTFLAGS);
-    WriteWorldMapStuff(dest + GAMESAVE_OFFSET_GMINFO, &gGMData);
+    WritePermanentFlags(dest->permanentFlags);
+    WriteWorldMapStuff(&dest->wmStuff, &gGMData);
     SaveDungeonRecords(dungeon);
-    WriteAndVerifySramFast(dungeon, dest + GAMESAVE_OFFSET_DUNGEON ,sizeof(dungeon));
+    WriteAndVerifySramFast(dungeon, dest->dungeons, sizeof(dungeon));
 
     chunk.magic32 = SAVEMAGIC32;
     chunk.kind = SAVEBLOCK_KIND_GAME;
@@ -206,7 +205,7 @@ void ReadGameSave(int slot)
 {
     int i;
     struct Dungeon dungeon[2];
-    void *src = GetSaveReadAddr(slot);
+    struct GameSaveBlock *src = GetSaveReadAddr(slot);
 
     if (!(PLAY_FLAG_HARD & gBmSt.gameStateBits))
         InvalidateSuspendSave(SAVE_ID_SUSPEND);
@@ -218,16 +217,16 @@ void ReadGameSave(int slot)
     InitUnits();
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++)
-        LoadSavedUnit(src + GAMESAVE_OFFSET_UNITS + i * sizeof(struct GameSavePackedUnit), &gUnitArrayBlue[i]);
+        LoadSavedUnit(src->units + i, &gUnitArrayBlue[i]);
 
-    ReadGMMonsterRnState(src + GAMESAVE_OFFSET_GMMONSTER);
-    ReadSupplyItems(src + GAMESAVE_OFFSET_SUPPLY);
-    ReadPermanentFlags(src + GAMESAVE_OFFSET_PERMANENTFLAGS);
-    ReadPidStats(src + GAMESAVE_OFFSET_PIDSTATS);
-    ReadChapterStats(src + GAMESAVE_OFFSET_CHAPTERSTATS);
+    ReadGMMonsterRnState(&src->gmUnit);
+    ReadSupplyItems(src->supplyItems);
+    ReadPermanentFlags(src->permanentFlags);
+    ReadPidStats(src->pidStats);
+    ReadChapterStats(src->chapterStats);
     ReadBonusContentClaimFlags(src);
-    ReadWorldMapStuff(src + GAMESAVE_OFFSET_GMINFO, &gGMData);
-    ReadSramFast(src + GAMESAVE_OFFSET_DUNGEON, dungeon, sizeof(dungeon));
+    ReadWorldMapStuff(&src->wmStuff, &gGMData);
+    ReadSramFast(src->dungeons, dungeon, sizeof(dungeon));
     LoadDungeonRecords(dungeon);
     WriteLastGameSaveId(slot);
 }
@@ -239,28 +238,28 @@ bool IsSaveValid(int index)
 
 void ReadGameSavePlaySt(int slot, struct PlaySt *buf)
 {
-    void *src = GetSaveReadAddr(slot);
-    ReadSramFast(src, buf, sizeof(struct PlaySt));
+    const struct GameSaveBlock *src = GetSaveReadAddr(slot);
+    ReadSramFast(&src->playSt, buf, sizeof(struct PlaySt));
 }
 
 u32 LoadSavedBonusClaimFlags(int slot)
 {
     u32 buf;
-    void *src = GetSaveReadAddr(slot);
-    ReadSramFast(src + GAMESAVE_OFFSET_BONUSCLAIM, &buf, sizeof(buf));
+    const struct GameSaveBlock *src = GetSaveReadAddr(slot);
+    ReadSramFast(&src->bonusClaimFlags, &buf, sizeof(buf));
     return buf;
 }
 
-void LoadSavedWMStuff(int slot, void *dest)
+void LoadSavedWMStuff(int slot, struct GMapData *dest)
 {
-    void *src = GetSaveReadAddr(slot);
-    ReadWorldMapStuff(src + GAMESAVE_OFFSET_GMINFO, dest);
+    const struct GameSaveBlock *src = GetSaveReadAddr(slot);
+    ReadWorldMapStuff(&src->wmStuff, dest);
 }
 
 s8 LoadSavedEid8A(int slot)
 {
-    void *sram_base = GetSaveReadAddr(slot);
-    ReadPermanentFlags_ret(sram_base + GAMESAVE_OFFSET_PERMANENTFLAGS, gGenericBuffer);
+    const struct GameSaveBlock *sram_base = GetSaveReadAddr(slot);
+    ReadPermanentFlags_ret(sram_base->permanentFlags, gGenericBuffer);
     return CheckPermanentFlagFrom(0x8A, gGenericBuffer);
 }
 
@@ -482,7 +481,7 @@ void InvalidateSuspendSave(int slot)
 
 void WriteSuspendSave(int slot)
 {
-    void *dest;
+    struct SuspendSaveBlock *dest;
     struct SaveBlockInfo chunk;
     u8 list[MENU_OVERRIDE_MAX];
     struct Dungeon dungeon[2];
@@ -498,15 +497,15 @@ void WriteSuspendSave(int slot)
     slot += GetNextSuspendSaveId();
     dest = GetSaveWriteAddr(slot);
     gPlaySt.time_saved = GetGameClock();
-    WriteAndVerifySramFast(&gPlaySt, dest + SUSPENDSAVE_OFFSET_PLAYST, sizeof(gPlaySt));
+    WriteAndVerifySramFast(&gPlaySt, &dest->playSt, sizeof(gPlaySt));
     StoreRNStateToActionStruct();
-    WriteAndVerifySramFast(&gActionData, dest + SUSPENDSAVE_OFFSET_ACTION, sizeof(struct ActionData));
+    WriteAndVerifySramFast(&gActionData, &dest->action, sizeof(struct ActionData));
 
     buf = (struct SuspendSavePackedUnit *)gGenericBuffer;
     for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++)
         EncodeSuspendSavePackedUnit(&gUnitArrayBlue[i], buf++);
 
-    WriteSramFast(gGenericBuffer, dest + SUSPENDSAVE_OFFSET_UNITS_BLUE, UNIT_SAVE_AMOUNT_BLUE * sizeof(struct SuspendSavePackedUnit));
+    WriteSramFast(gGenericBuffer, (u8 *)dest->blueUnits, UNIT_SAVE_AMOUNT_BLUE * sizeof(struct SuspendSavePackedUnit));
 
     buf = (struct SuspendSavePackedUnit *)gGenericBuffer;
     for (i = 0; i < UNIT_SAVE_AMOUNT_RED; i++)
@@ -515,31 +514,31 @@ void WriteSuspendSave(int slot)
     for (i = 0; i < UNIT_SAVE_AMOUNT_GREEN; i++)
         EncodeSuspendSavePackedUnit(&gUnitArrayGreen[i], buf++);
 
-    WriteSramFast(gGenericBuffer, dest + SUSPENDSAVE_OFFSET_UNITS_RED, (UNIT_SAVE_AMOUNT_RED + UNIT_SAVE_AMOUNT_GREEN) * sizeof(struct SuspendSavePackedUnit));
+    WriteSramFast(gGenericBuffer, (u8 *)dest->redUnits, (UNIT_SAVE_AMOUNT_RED + UNIT_SAVE_AMOUNT_GREEN) * sizeof(struct SuspendSavePackedUnit)); // TODO: maybe pack them into a single struct?
 
-    WritePermanentFlags(dest + SUSPENDSAVE_OFFSET_PERMANENTFLAGS);
-    WriteChapterFlags(dest + SUSPENDSAVE_OFFSET_CHAPTERFLAGS);
-    WriteSupplyItems(dest + SUSPENDSAVE_OFFSET_SUPPLY);
-    WritePidStats(dest + SUSPENDSAVE_OFFSET_PIDSTATS);
-    WriteChapterStats(dest + SUSPENDSAVE_OFFSET_CHAPTERSTATS);
-    WriteTraps(dest + SUSPENDSAVE_OFFSET_TRAPS);
+    WritePermanentFlags(dest->permanentFlags);
+    WriteChapterFlags(dest->chapterFlags);
+    WriteSupplyItems(dest->supplyItems);
+    WritePidStats(dest->pidStats);
+    WriteChapterStats(dest->chapterStats);
+    WriteTraps(dest->traps);
 
     GetForceDisabledMenuItems(list);
-    WriteAndVerifySramFast(list, dest + SUSPENDSAVE_OFFSET_FORCENMENU, sizeof(list));
+    WriteAndVerifySramFast(list, dest->menuOverride, sizeof(list));
 
-    WriteWorldMapStuff(dest + 0x1F44, &gGMData);
+    WriteWorldMapStuff(&dest->wmStuff, &gGMData);
 
     /**
-     * I think this maybe a coincidence that I$ find that gm-generated-monster does not require the suspend save sapce
+     * I think this maybe a coincidence that I$ find that gm-generated-monster does not require the suspend save space
      */
     SaveDungeonRecords(dungeon);
-    WriteAndVerifySramFast(dungeon, dest + SUSPENDSAVE_OFFSET_GMMONSTER, 2 * sizeof(struct Dungeon));
+    WriteAndVerifySramFast(dungeon, &dest->wmMonsterUnit, sizeof(dungeon));
 
     SaveDungeonState(dungeon);
-    WriteAndVerifySramFast(dungeon, dest + SUSPENDSAVE_OFFSET_DUNGEON, sizeof(struct Dungeon));
+    WriteAndVerifySramFast(dungeon, &dest->dungeon, sizeof(struct Dungeon));
 
     val = GetEventSlotCounter();
-    WriteAndVerifySramFast(&val, dest + SUSPENDSAVE_OFFSET_EVENTSLOTCNT, sizeof(int));
+    WriteAndVerifySramFast(&val, &dest->eventSlotCnt, sizeof(int));
 
     chunk.magic32 = SAVEMAGIC32;
     chunk.kind = SAVEBLOCK_KIND_SUSPEND;
@@ -554,43 +553,43 @@ void ReadSuspendSave(int slot)
     int i, val;
     u8 list[MENU_OVERRIDE_MAX];
     struct Dungeon dungeon[2];
-    void *src = GetSaveReadAddr(slot + gSuspendSaveIdOffset);
+    struct SuspendSaveBlock *src = GetSaveReadAddr(slot + gSuspendSaveIdOffset);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_PLAYST, &gPlaySt, sizeof(gPlaySt));
+    ReadSramFast(&src->playSt, &gPlaySt, sizeof(gPlaySt));
     SetGameTime(gPlaySt.time_saved);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_ACTION, &gActionData, sizeof(struct ActionData));
+    ReadSramFast(&src->action, &gActionData, sizeof(struct ActionData));
     LoadRNStateFromActionStruct();
     InitUnits();
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++)
-        ReadSuspendSavePackedUnit(src + SUSPENDSAVE_OFFSET_UNITS_BLUE + i * sizeof(struct SuspendSavePackedUnit), &gUnitArrayBlue[i]);
+        ReadSuspendSavePackedUnit(&src->blueUnits[i], &gUnitArrayBlue[i]);
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_RED; i++)
-        ReadSuspendSavePackedUnit(src + SUSPENDSAVE_OFFSET_UNITS_RED + i * sizeof(struct SuspendSavePackedUnit), &gUnitArrayRed[i]);
+        ReadSuspendSavePackedUnit(&src->redUnits[i], &gUnitArrayRed[i]);
 
     for (i = 0; i < UNIT_SAVE_AMOUNT_GREEN; i++)
-        ReadSuspendSavePackedUnit(src + SUSPENDSAVE_OFFSET_UNITS_GREEN + i * sizeof(struct SuspendSavePackedUnit), &gUnitArrayGreen[i]);
+        ReadSuspendSavePackedUnit(&src->greenUnits[i], &gUnitArrayGreen[i]);
 
-    ReadPidStats(src + SUSPENDSAVE_OFFSET_PIDSTATS);
-    ReadChapterStats(src + SUSPENDSAVE_OFFSET_CHAPTERSTATS);
-    ReadSupplyItems(src + SUSPENDSAVE_OFFSET_SUPPLY);
-    ReadPermanentFlags(src + SUSPENDSAVE_OFFSET_PERMANENTFLAGS);
-    ReadChapterFlags(src + SUSPENDSAVE_OFFSET_CHAPTERFLAGS);
-    ReadTraps(src + SUSPENDSAVE_OFFSET_TRAPS);
+    ReadPidStats(src->pidStats);
+    ReadChapterStats(src->chapterStats);
+    ReadSupplyItems(src->supplyItems);
+    ReadPermanentFlags(src->permanentFlags);
+    ReadChapterFlags(src->chapterFlags);
+    ReadTraps(src->traps);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_FORCENMENU, list, sizeof(list));
+    ReadSramFast(src->menuOverride, list, sizeof(list));
     SetForceDisabledMenuItems(list);
 
-    ReadWorldMapStuff(src + SUSPENDSAVE_OFFSET_GMINFO, &gGMData);
+    ReadWorldMapStuff(&src->wmStuff, &gGMData);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_GMMONSTER, dungeon, 2 * sizeof(struct Dungeon));
+    ReadSramFast(&src->wmMonsterUnit, dungeon, sizeof(dungeon));
     LoadDungeonRecords(dungeon);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_DUNGEON, dungeon, sizeof(struct Dungeon));
+    ReadSramFast(&src->dungeon, dungeon, sizeof(struct Dungeon));
     LoadDungeonState(dungeon);
 
-    ReadSramFast(src + SUSPENDSAVE_OFFSET_EVENTSLOTCNT, &val, sizeof(int));
+    ReadSramFast(&src->eventSlotCnt, &val, sizeof(int));
     SetEventSlotCounter(val);
 
     SetBonusContentClaimFlags(LoadSavedBonusClaimFlags(gPlaySt.gameSaveSlot));

--- a/src/gamerankings.c
+++ b/src/gamerankings.c
@@ -454,10 +454,10 @@ int GetChapterFundsRank(void) {
 
     int totalGold = GetPartyTotalGoldValue();
 
-    goldInChapter = gPlaySt.total_gold;
+    goldInChapter = gPlaySt.unk_30.total_gold;
     goldInChapter = totalGold - goldInChapter;
 
-    gPlaySt.total_gold = totalGold;
+    gPlaySt.unk_30.total_gold = totalGold;
 
     ent = GetChapterStats(GetNextChapterStatsSlot() - 1);
 
@@ -495,14 +495,14 @@ int GetChapterWinPerc(void) {
         totalWins = 0xFFFFF;
     }
 
-    chapterTotalBattles = gPlaySt.unk_34_00;
+    chapterTotalBattles = gPlaySt.unk_30.unk_4_00;
 
     if (totalBattles == chapterTotalBattles) {
         return 40;
     }
 
-    a = gPlaySt.unk_34_14;
-    b = gPlaySt.unk_38_1 << 12;
+    a = gPlaySt.unk_30.unk_4_14;
+    b = gPlaySt.unk_30.unk_8_1 << 12;
     num = (totalWins - (b | a)) * 100;
 
     percentage = num / (totalBattles - chapterTotalBattles);
@@ -511,9 +511,9 @@ int GetChapterWinPerc(void) {
         percentage = 100;
     }
 
-    gPlaySt.unk_34_00 = totalBattles;
-    gPlaySt.unk_34_14 = (totalWins & 0x00000FFF);
-    gPlaySt.unk_38_1 = ((u32)totalWins >> 12);
+    gPlaySt.unk_30.unk_4_00 = totalBattles;
+    gPlaySt.unk_30.unk_4_14 = (totalWins & 0x00000FFF);
+    gPlaySt.unk_30.unk_8_1 = ((u32)totalWins >> 12);
 
     return percentage;
 }
@@ -559,8 +559,8 @@ int GetChapterExpRank(void) {
         totalExp = 0xFFFFF;
     }
 
-    expInChapter = totalExp - gPlaySt.unk_38_2;
-    gPlaySt.unk_38_2 = totalExp;
+    expInChapter = totalExp - gPlaySt.unk_30.unk_8_2;
+    gPlaySt.unk_30.unk_8_2 = totalExp;
 
     ent = GetChapterStats(GetNextChapterStatsSlot() - 1);
 
@@ -592,19 +592,19 @@ void ComputeChapterRankings(void) {
             case CHAPTER_MODE_COMMON:
             case CHAPTER_MODE_EIRIKA:
             case CHAPTER_MODE_EPHRAIM:
-                gPlaySt.tacticsRank = GetChapterTacticsRank();
-                gPlaySt.survivalRank = GetChapterSurvivalRank();
-                gPlaySt.fundsRank = GetChapterFundsRank();
-                gPlaySt.combatRank = GetChapterCombatRank();
-                gPlaySt.expRank = GetChapterExpRank();
+                gPlaySt.unk_30.tacticsRank = GetChapterTacticsRank();
+                gPlaySt.unk_30.survivalRank = GetChapterSurvivalRank();
+                gPlaySt.unk_30.fundsRank = GetChapterFundsRank();
+                gPlaySt.unk_30.combatRank = GetChapterCombatRank();
+                gPlaySt.unk_30.expRank = GetChapterExpRank();
         }
 
         overallRank = GetOverallRank(
-            gPlaySt.tacticsRank,
-            gPlaySt.survivalRank,
-            gPlaySt.fundsRank,
-            gPlaySt.expRank,
-            gPlaySt.combatRank
+            gPlaySt.unk_30.tacticsRank,
+            gPlaySt.unk_30.survivalRank,
+            gPlaySt.unk_30.fundsRank,
+            gPlaySt.unk_30.expRank,
+            gPlaySt.unk_30.combatRank
         );
 
         newRank = gPlaySt.unk_2C_04 + overallRank;

--- a/src/savemenu.c
+++ b/src/savemenu.c
@@ -128,20 +128,15 @@ void sub_80A88E0(struct SaveMenuProc* proc) {
     return;
 }
 
-#if NONMATCHING
-
-/* https://decomp.me/scratch/5PCbw */
-
 //! FE8U = 0x080A8950
 int LoadSaveMenuHelpText(int slot) {
     int leaderId;
-    uintptr_t saveBase;
+    struct GameSaveBlock *saveBase;
     int i;
     struct PlaySt chapterData;
     struct Unit unit;
     struct GMapData mapData;
-
-    u8 localbuffer[0x30];
+    u8 localbuffer[4] __attribute__((unused));
 
     if (!IsSaveValid(slot)) {
         return 0;
@@ -162,119 +157,22 @@ int LoadSaveMenuHelpText(int slot) {
 
     saveBase = GetSaveReadAddr(slot);
 
-    for (i = 0;; i++) {
-
-        if (i >= 0x33) {
-            sub_80AA700();
-            return 2;
-        }
-
-        LoadSavedUnit((void*)((saveBase + 0x4c) + 0x24 * i), &unit);
-
-        if (unit.pCharacterData == NULL) {
-            continue;
-        }
-
-        if (unit.pCharacterData->number != leaderId) {
-            continue;
-        }
-
-        gUnknown_0203EF64.unk_00 = leaderId;
-        gUnknown_0203EF64.unk_01 = unit.level;
-
-        ReadWorldMapStuff((void*)(saveBase + 0xD8C), &mapData);
-        gUnknown_0203EF64.unk_02 = mapData.unk10[0].location;
-
-        return 2;
+    for (i = 0; i < UNIT_SAVE_AMOUNT_BLUE; i++) {
+        LoadSavedUnit(&saveBase->units[i], &unit);
+        if (unit.pCharacterData != NULL && unit.pCharacterData->number == leaderId) break;
     }
+    if (i >= UNIT_SAVE_AMOUNT_BLUE) goto label;
+    gUnknown_0203EF64.unk_00 = leaderId;
+    gUnknown_0203EF64.unk_01 = unit.level;
 
+    ReadWorldMapStuff(&saveBase->wmStuff, &mapData);
+    gUnknown_0203EF64.unk_02 = mapData.unk10[0].location;
+
+    return 2;
+label:
+    sub_80AA700();
+    return 2;
 }
-
-#else // if !NONMATCHING
-
-__attribute__((naked))
-int LoadSaveMenuHelpText(int slot) {
-    asm("\n\
-        .syntax unified\n\
-        push {r4, r5, r6, r7, lr}\n\
-        sub sp, #0x168\n\
-        adds r4, r0, #0\n\
-        bl IsSaveValid\n\
-        lsls r0, r0, #0x18\n\
-        cmp r0, #0\n\
-        bne _080A8964\n\
-        movs r0, #0\n\
-        b _080A89DA\n\
-    _080A8964:\n\
-        adds r0, r4, #0\n\
-        mov r1, sp\n\
-        bl ReadGameSavePlaySt\n\
-        mov r0, sp\n\
-        ldrb r0, [r0, #0x1b]\n\
-        cmp r0, #1\n\
-        blt _080A897C\n\
-        cmp r0, #2\n\
-        ble _080A897C\n\
-        cmp r0, #3\n\
-        beq _080A8980\n\
-    _080A897C:\n\
-        movs r6, #1\n\
-        b _080A8982\n\
-    _080A8980:\n\
-        movs r6, #0xf\n\
-    _080A8982:\n\
-        adds r0, r4, #0\n\
-        bl GetSaveReadAddr\n\
-        adds r7, r0, #0\n\
-        movs r5, #0\n\
-        adds r4, r7, #0\n\
-        adds r4, #0x4c\n\
-        b _080A8996\n\
-    _080A8992:\n\
-        adds r4, #0x24\n\
-        adds r5, #1\n\
-    _080A8996:\n\
-        cmp r5, #0x32\n\
-        bgt _080A89D4\n\
-        adds r0, r4, #0\n\
-        add r1, sp, #0x4c\n\
-        bl LoadSavedUnit\n\
-        ldr r0, [sp, #0x4c]\n\
-        add r1, sp, #0x4c\n\
-        cmp r0, #0\n\
-        beq _080A8992\n\
-        ldrb r0, [r0, #4]\n\
-        cmp r0, r6\n\
-        bne _080A8992\n\
-        ldr r4, _080A89CC  @ gUnknown_0203EF64\n\
-        strb r6, [r4]\n\
-        ldrb r0, [r1, #8]\n\
-        strb r0, [r4, #1]\n\
-        ldr r1, _080A89D0  @ 0x00000D8C\n\
-        adds r0, r7, r1\n\
-        add r5, sp, #0x94\n\
-        adds r1, r5, #0\n\
-        bl ReadWorldMapStuff\n\
-        ldrb r0, [r5, #0x11]\n\
-        strb r0, [r4, #2]\n\
-        b _080A89D8\n\
-        .align 2, 0\n\
-    _080A89CC: .4byte gUnknown_0203EF64\n\
-    _080A89D0: .4byte 0x00000D8C\n\
-    _080A89D4:\n\
-        bl sub_80AA700\n\
-    _080A89D8:\n\
-        movs r0, #2\n\
-    _080A89DA:\n\
-        add sp, #0x168\n\
-        pop {r4, r5, r6, r7}\n\
-        pop {r1}\n\
-        bx r1\n\
-        .syntax divided\n\
-    ");
-}
-
-#endif
 
 //! FE8U = 0x080A89E4
 s8 sub_80A89E4(struct SaveMenuProc* proc) {

--- a/src/worldmap.c
+++ b/src/worldmap.c
@@ -3,11 +3,10 @@
 #include "functions.h"
 
 //! FE8U = 0x080BC8BC
-bool SetupNewWMRoute(struct GMapData *pGMapData, u8 *routeData, int idx) {
-    u8 *pNumRoutes = routeData + 0x20;
-    if (*pNumRoutes < 0x20) {
-        routeData[(s8) *pNumRoutes] = (u8) idx;
-        *pNumRoutes += 1;
+bool SetupNewWMRoute(struct GMapData *pGMapData, struct OpenPaths *routeData, int idx) {
+    if (routeData->openPathsLength < sizeof(routeData->openPaths)) {
+        routeData->openPaths[routeData->openPathsLength] = idx;
+        routeData->openPathsLength += 1;
         sub_80BCA0C(pGMapData);
         return false;
     }

